File: apps/flutter/lib/features/csv_import/domain/entities/import_job.dart
class ImportJob {
  final String id;
  final String userId;
  final String status;
  final String accountId;
  final String bankType;
  final int totalRows;
  final int imported;
  final int duplicates;
  final int errors;
  final List<ImportErrorDetail> errorDetails;
  final DateTime createdAt;
  final DateTime? completedAt;

  const ImportJob({
    required this.id,
    required this.userId,
    required this.status,
    required this.accountId,
    required this.bankType,
    required this.totalRows,
    required this.imported,
    required this.duplicates,
    required this.errors,
    required this.errorDetails,
    required this.createdAt,
    this.completedAt,
  });

  factory ImportJob.fromJson(Map<String, dynamic> json) {
    return ImportJob(
      id: (json["importJobId"] ?? json["id"]) as String,
      userId: json["userId"] as String,
      status: json["status"] as String,
      accountId: json["accountId"] as String,
      bankType: json["bankType"] as String,
      totalRows: (json["totalRows"] as num?)?.toInt() ?? 0,
      imported: (json["imported"] as num?)?.toInt() ?? 0,
      duplicates: (json["duplicates"] as num?)?.toInt() ?? 0,
      errors: (json["errors"] as num?)?.toInt() ?? 0,
      errorDetails: (json["errorDetails"] as List<dynamic>?)
              ?.map((item) => ImportErrorDetail.fromJson(item as Map<String, dynamic>))
              .toList() ??
          const [],
      createdAt: DateTime.parse(json["createdAt"] as String),
      completedAt: json["completedAt"] != null
          ? DateTime.tryParse(json["completedAt"] as String)
          : null,
    );
  }
}

class ImportErrorDetail {
  final int row;
  final String error;

  const ImportErrorDetail({
    required this.row,
    required this.error,
  });

  factory ImportErrorDetail.fromJson(Map<String, dynamic> json) {
    return ImportErrorDetail(
      row: (json["row"] as num?)?.toInt() ?? 0,
      error: json["error"] as String,
    );
  }
}

File: packages/database/src/models/import_job.ts
import { AggregateRoot } from "@abejarano/ts-mongodb-criteria"
import { BankType } from "./bank_type"
import { createMongoId } from "./shared/mongo_id"

export enum ImportJobStatus {
  Pending = "pending",
  Processing = "processing",
  Completed = "completed",
  Failed = "failed",
}

export type ImportJobPrimitives = {
  id?: string
  importJobId: string
  userId: string
  status: ImportJobStatus
  accountId: string
  bankType: BankType
  totalRows: number
  imported: number
  duplicates: number
  errors: number
  errorDetails: Array<{ row: number; error: string }>
  createdAt: Date
  completedAt?: Date | null
}

export type ImportJobCreateProps = {
  userId: string
  accountId: string
  bankType: BankType
  totalRows: number
}

export class ImportJob extends AggregateRoot {
  private readonly props: ImportJobPrimitives

  private constructor(props: ImportJobPrimitives) {
    super()
    this.props = props
  }

  static create(props: ImportJobCreateProps): ImportJob {
    const now = new Date()

    return new ImportJob({
      importJobId: createMongoId(),
      userId: props.userId,
      status: ImportJobStatus.Pending,
      accountId: props.accountId,
      bankType: props.bankType,
      totalRows: props.totalRows,
      imported: 0,
      duplicates: 0,
      errors: 0,
      errorDetails: [],
      createdAt: now,
      completedAt: null,
    })
  }

  getId(): string {
    return this.props.id ?? this.props.importJobId
  }

  getImportJobId(): string {
    return this.props.importJobId
  }

  toPrimitives(): ImportJobPrimitives {
    return this.props
  }

  static override fromPrimitives(primitives: ImportJobPrimitives): ImportJob {
    return new ImportJob(primitives)
  }
}
